crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+                                        pubcast(bits, p) : \
+                                        &(prvcast(bits, q)->pub))
+# define constish(bits, p, q)   pubconst(bits, (p) != NULL ? \
+                                         pubconst(bits, p) : \
+                                         &(prvconst(bits, q)->pub))
+
+int ossl_ml_kem_vencode_public_key(vinfo_t v,
+                                   uint8_t *out,
+                                   size_t len,
+                                   const void *pub,
+                                   const void *prv)
+{
+# define case_encode_pubkey(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if ((pub == NULL && prv == NULL) || len != v->pubkey_bytes) \
+            return 0; \
+        encode_pubkey(bits)(out, constish(bits, pub, prv)); \
+        return 1
+
+    switch (v->rank) {
+    case_encode_pubkey(512);
+    case_encode_pubkey(768);
+    case_encode_pubkey(1024);
+    }
+    return 0;
+# undef case_encode_pubkey
+}
+
+int ossl_ml_kem_vparse_public_key(vinfo_t  v,
+                                  void **pub,
+                                  const uint8_t *in,
+                                  size_t len,
+                                  mctx *ctx)
+{
+# define case_parse_pubkey(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if (pub == NULL || len != v->pubkey_bytes) \
+            return 0; \
+        if (*pub == NULL \
+            && (*pub = OPENSSL_malloc(sizeof(*pubcast(bits, 0)))) == NULL) \
+            return 0; \
+        return parse_pubkey(bits)(pubcast(bits, *pub), in, ctx)
+
+    switch (v->rank) {
+    case_parse_pubkey(512);
+    case_parse_pubkey(768);
+    case_parse_pubkey(1024);
+    }
+    return 0;
+# undef case_parse_pubkey
+}
+
+int ossl_ml_kem_vencode_private_key(vinfo_t v,
+                                    uint8_t *out,
+                                    size_t len,
+                                    const void *prv)
+{
+# define case_encode_prvkey(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if (prv == NULL || len != v->prvkey_bytes) \
+            return 0; \
+        encode_prvkey(bits)(out, prvconst(bits, prv)); \
+        return 1
+
+    switch (v->rank) {
+    case_encode_prvkey(512);
+    case_encode_prvkey(768);
+    case_encode_prvkey(1024);
+    }
+    return 0;
+# undef case_encode_prvkey
+}
+
+int ossl_ml_kem_vparse_private_key(vinfo_t v,
+                                   void **prv,
+                                   const uint8_t *in,
+                                   size_t len,
+                                   mctx *ctx)
+{
+# define case_parse_prvkey(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if (prv == NULL || len != v->prvkey_bytes) \
+            return 0; \
+        if (*prv == NULL \
+            && (*prv = OPENSSL_malloc(sizeof(*prvcast(bits, 0)))) == NULL) \
+            return 0; \
+        return parse_prvkey(bits)(prvcast(bits, *prv), in, ctx)
```

Comment:
on failure of parse_prvkey what level is responsible for freeing *prv?
Normally thjis would be cleaned up at the same level.

Reply:
Fixed in #26172